### ISSUE TITLE
refactor: add IndexManager class with batch-wise index population

### ISSUE THREAD

**Issue by goergen95** [goergen95/seapig#79]
This document contains the phased implementation plan and the design for an IndexManager class that centralises all index-related functionality used by KNNScore and its derived classes. The revised plan prioritises HNSW (Hierarchical Navigable Small World) indexes via nmslib as the default approximate nearest-neighbour backend while retaining a robust brute-force fallback and supporting both full and batch-wise index population.

Goals

- Extract index build/search/load/save responsibilities from KNNScore into a single well-typed class.
- Keep KNNScore focused on score computation and statistics (k, stat, distance type).
- Make HNSW via nmslib the primary approximate index option (configurable).
- Support both batch-wise population of indexes and building from all reference embeddings at once.
- Preserve existing public KNNScore API where possible.
- Add unit tests covering IndexManager behaviour before integrating it into KNNScore.

Design: IndexManager (overview)

Responsibilities

- Hold and validate reference and optional validation embeddings (torch.Tensor).
- Build a nearest-neighbour index (nmslib HNSW-backed by default) with optional preprocessing (PCA/whitening).
- Allow adding reference vectors in batches (incremental population) or providing the full reference set for single-shot index creation.
- Provide search/nearest-neighbor interfaces returning indices and/or distances as torch.Tensor.
- Persist and restore index and related metadata (embeddings, PCA params, HNSW params) to disk.
- Provide utilities to compute Mahalanobis components (covariance, precision / inverse) when requested.
- Manage device/dtype placement for tensors used in search results.

Public API (suggested)

- class IndexManager:
  - __init__(
      self,
      method: str = "hnsw",  # "hnsw" (nmslib) | "brute"
      device: str | None = None,
      pca_components: int | None = None,
      pca_exp_var: float | None = None,
      dtype: "torch.dtype" = "torch.float32",
      logger: "logging.Logger" | None = None,
  ) -> None
  - fit(self, ref_embeddings: "torch.Tensor" | None = None, cal_embeddings: "torch.Tensor" | None = None) -> None
    - Attach reference embeddings in-memory (optional). Calling fit without ref_embeddings is
      allowed when the user intends to populate the index incrementally via add_batch().
  - add_batch(self, embeddings: "torch.Tensor", ids: "torch.Tensor" | None = None) -> None
    - Add a batch of reference vectors to the internal store and (if index exists) to the index.
  - build_index(self, from_memory: bool = True, hnsw_params: dict | None = None, **nmslib_options) -> None
    - Build or (re)build the underlying index. If from_memory is True the builder will use
      all stored reference embeddings; otherwise the implementation may accept an external
      data source. hnsw_params should expose M, efConstruction, and other HNSW options.
  - search(self, queries: "torch.Tensor", k: int, return_distances: bool = True) -> tuple["torch.Tensor", "torch.Tensor"] | "torch.Tensor"
  - get_ref_embeddings(self) -> "torch.Tensor" | None
  - get_val_embeddings(self) -> "torch.Tensor" | None
  - save(self, path: str) -> dict[str, str]
  - load(self, path: str) -> None
  - reset(self) -> None

Design notes

- All public methods will be fully typed and documented with NumPy-style docstrings.
- NMSLIB HNSW integration is the default approximate backend; method="brute" falls back to a pure-PyTorch/NumPy brute-force search.
- HNSW-specific configuration (M, efConstruction, efSearch) must be exposed via hnsw_params and documented in the API.
- Support both modes of population:
  1. Batch-wise population: call add_batch() repeatedly; build_index(from_memory=True) will construct from accumulated vectors or add vectors to an existing nmslib index incrementally if supported.
  2. Single-shot full build: pass the full reference tensor to fit() or ensure all vectors are added then call build_index(from_memory=True).
- Search should accept batched queries and search option parameters and preserve the input device/dtype in returned tensors.
- Persistence: save embeddings as torch tensors (torch.save). For nmslib/HNSW save the index using nmslib.saveIndex; save metadata (method, dtype, PCA info, HNSW params) as JSON alongside. For brute-force store necessary metadata/pickled helpers.
- PCA: If pca_components or pca_exp_var provided, compute PCA (via scikit PCA/IncrementalPCA) during fit/add_batch and store its state_dict; the index stores reduced-dimension embeddings.
- Mahalanobis: provide utilities to compute covariance and precision from stored reference embeddings; allow pre-transforming embeddings before inserting into the index.
- Thread-safety: class is not required to be thread-safe initially; document this.
- Logging: use injected logger or getLogger(__name__). Use debug/info for major events (fit, add_batch, build_index, search, save, load).

HNSW-specific considerations

- Prefer HNSW (nmslib) for medium-to-large embedding sets because it balances query speed and build time.
- Document recommended defaults: M=16, efConstruction=200, efSearch configurable per-query.
- Provide a clear fallback path when nmslib is unavailable (method="brute").

Testing matrix (unit tests)

- test_indexmanager_fit_and_search_brute: verify exact nearest neighbours with brute force for small datasets.
- test_indexmanager_nmslib_hnsw_build_and_search: build HNSW index and assert returned indices/distances match brute force within tolerance; test with multiple hnsw_params.
- test_indexmanager_nmslib_batch_add_and_search: add reference vectors in multiple batches, build index (or add directly), and assert search results match single-shot build.
- test_indexmanager_persistence: save and load an index (HNSW and brute) and assert searches give same results.
- test_indexmanager_pca_reduce: verify PCA transforms dimension and search uses reduced space.
- test_indexmanager_input_validation: invalid input shapes, mismatched dtypes cause clear exceptions.

Phased implementation plan

Phase 0 — Preparation (small)
- Add plan file (this document).
- Add placeholder module: seapig/scores/index_manager.py containing the class skeleton and full docstrings (no detailed implementation).
- Add tests scaffolding file tests/test_index_manager.py with empty test functions (marked xfail or skipped) to outline expectations.

Acceptance for Phase 0: file and tests present; CI green for linting/type checks.

Phase 1 — Implement IndexManager core (brute-force + persistence) and tests
- Implement fit(), get_ref_embeddings(), search() using a brute-force torch implementation.
- Implement batch storage semantics and add_batch() that appends to an internal buffer and optionally to an existing brute index structure.
- Implement save()/load() for embeddings and metadata (torch.save + json metadata).
- Add input validation and docstrings + type annotations.
- Implement unit tests: fit_and_search_brute, persistence, input validation.

Acceptance for Phase 1: unit tests for brute-force pass locally and in CI. KNNScore not modified yet.

Phase 2 — NMSLIB HNSW integration and PCA
- Add nmslib HNSW-backed index creation in build_index(), with hnsw_params exposed to callers.
- Implement incremental population for nmslib where feasible (use addDataPointBatch or equivalent); ensure add_batch() works for both adding to memory and directly to an existing nmslib index.
- Integrate PCA preprocessing (pca_components / pca_exp_var) into fit() / add_batch() and index building.
- Ensure search() abstracts over underlying index (brute vs nmslib).
- Add tests for nmslib HNSW, batch population, and PCA.

Acceptance for Phase 2: NMSLIB-backed tests pass in environments with nmslib installed; PCA tests validate dimensionality and search correctness.

Phase 3 — KNNScore integration
- Refactor seapig/scores/knn.py: replace internal index handling with an IndexManager instance. Start with KNNScore._setup_index() using IndexManager.fit()/build_index().
- Update derived classes (Cosine, Euclidean, MahalanobisScore) to call IndexManager.search() and use returned distances.
- Keep backward-compatible behaviour: maintain KNNScore public API but delegate index tasks.
- Add integration tests: run existing test_knn.py with the new refactor; adjust tests where necessary.

Acceptance for Phase 3: tests/test_knn.py pass with new IndexManager-based implementation.

Phase 4 — Clean up, docs, and deprecation
- Remove duplicated index helper functions from knn.py.
- Add detailed docstrings for IndexManager in the API docs and a short guide in docs/ describing migration.
- If any public behaviour changed, add deprecation warnings in KNNScore and schedule full removal in future major release.
- Add performance tests and benchmarks if desired.

Acceptance for Phase 4: repository tests and style checks pass. Documentation updated.

Migration checklist for maintainers

- Run tests before modifying KNNScore to have a baseline.
- Implement Phase 1 and land tests.
- Implement Phase 2 and rerun CI on platforms with nmslib.
- Implement Phase 3, doing small commits with clear messages (e.g., "refactor(knn): delegate index to IndexManager").
- Keep PRs small and request review for API changes.

Notes and trade-offs

- Design the IndexManager API to be flexible for future extensions (e.g., adding new index types, distance metrics) without breaking existing code.
- Storing large embeddings to disk can be heavy; persistence should allow streaming/sparse formats if future need arises.
- PCA implementation choice: scikit-learn vs torch.svd. Prefer scikit-learn in CPU contexts for stability; provide a torch fallback.

Implementation hints and non-normative guidance

- Brute-force pairwise distances with torch.matmul and norms for cosine / Euclidean.
- When using nmslib HNSW, expose M, efConstruction, efSearch in hnsw_params and document recommended defaults (M=16, efConstruction=200).
- For incremental population, prefer adding to an in-memory buffer and calling build_index(from_memory=True) to ensure deterministic index state; support direct incremental adds to nmslib indexes where it is stable and reproducible.
- Use space="l2" for Euclidean; for cosine, normalise embeddings and use space="cosinesimil".
- For Mahalanobis, compute precision/vi and transform embeddings before inserting into the index when Mahalanobis scoring is used.
